File: packages/flutter-sdk/lib/src/connection_state.dart
enum FluxyChatTransport { websocket, sse, polling, none }
enum FluxyConnectionStateStatus { connected, connecting, reconnecting, polling, sse, disconnected }

class FluxyConnectionState {
  final FluxyConnectionStateStatus status;
  final Object? lastError;
  final int retryAttempt;
  final DateTime? nextRetryAt;
  final FluxyChatTransport transport;
  FluxyConnectionState({required this.status, this.lastError, this.retryAttempt = 0, this.nextRetryAt, required this.transport});
}

FluxyConnectionState buildFluxyConnectionState({
  required FluxyConnectionStateStatus status,
  Object? lastError,
  int retryAttempt = 0,
  int? reconnectDelayMs,
  FluxyChatTransport? transport,
}) {
  DateTime? nextRetryAt;
  if (status == FluxyConnectionStateStatus.reconnecting) {
    final delay = reconnectDelayMs ?? computeReconnectBackoffMs(retryAttempt > 0 ? retryAttempt : 1);
    if (delay > 0) nextRetryAt = DateTime.now().add(Duration(milliseconds: delay));
  }
  return FluxyConnectionState(
    status: status,
    lastError: lastError,
    retryAttempt: retryAttempt,
    nextRetryAt: nextRetryAt,
    transport: transport ?? _transportFromStatus(status),
  );
}

FluxyChatTransport _transportFromStatus(FluxyConnectionStateStatus status) {
  switch (status) {
    case FluxyConnectionStateStatus.sse: return FluxyChatTransport.sse;
    case FluxyConnectionStateStatus.polling: return FluxyChatTransport.polling;
    case FluxyConnectionStateStatus.connected:
    case FluxyConnectionStateStatus.connecting:
    case FluxyConnectionStateStatus.reconnecting: return FluxyChatTransport.websocket;
    default: return FluxyChatTransport.none;
  }
}

File: packages/react-native-sdk/src/connection-state.ts
import { computeReconnectBackoffMs } from './errors';

export type FluxyChatTransport = 'websocket' | 'sse' | 'polling' | 'none';
export type FluxyConnectionStateStatus = 'connected' | 'connecting' | 'reconnecting' | 'polling' | 'sse' | 'disconnected';

export interface FluxyConnectionState {
  status: FluxyConnectionStateStatus;
  lastError: Error | null;
  retryAttempt: number;
  nextRetryAt: string | null;
  transport: FluxyChatTransport;
}

export interface BuildFluxyConnectionStateInput {
  status: FluxyConnectionStateStatus;
  lastError?: Error | null;
  retryAttempt?: number;
  reconnectDelayMs?: number | null;
  transport?: FluxyChatTransport;
  nowMs?: number;
}

export function buildFluxyConnectionState(input: BuildFluxyConnectionStateInput): FluxyConnectionState {
  const nowMs = input.nowMs ?? Date.now();
  const retryAttempt = input.retryAttempt ?? 0;
  let nextRetryAt: string | null = null;

  if (input.status === 'reconnecting') {
    const delay = input.reconnectDelayMs ?? computeReconnectBackoffMs(Math.max(retryAttempt, 1));
    if (delay > 0) nextRetryAt = new Date(nowMs + delay).toISOString();
  }

  return {
    status: input.status,
    lastError: input.lastError ?? null,
    retryAttempt,
    nextRetryAt,
    transport: input.transport ?? transportFromStatus(input.status),
  };
}

function transportFromStatus(status: FluxyConnectionStateStatus): FluxyChatTransport {
  if (status === 'sse') return 'sse';
  if (status === 'polling') return 'polling';
  if (status === 'connected' || status === 'connecting' || status === 'reconnecting') return 'websocket';
  return 'none';
}
